apps/src/code-studio/pd/form_components_func/FormController.jsx

```diff
@@ -111,20 +120,31 @@ const FormController = props => {
   }, []);
 
   useEffect(() => {
-    // this function needs to be recreated because it holds 'submitting' in its closure
+    if (
+      !isEqual(
+        omit(data, autoComputedFields),
+        omit(savedData, autoComputedFields)
+      )
+    ) {
+      setHasUserChangedData(true);
+    } else {
+      setHasUserChangedData(false);
```

Comment:
calling `setHasUserChangedData` won't immediately change the value of `hasUserChangedData` later on in this function (it has to rerender first). I might try moving the set into a separate `useEffect` and have the eventListener stuff depend on `hasUserChangedData`

Reply:
Ah, good call. That was my original plan and then wanted to avoid two useEffects, but you're right––I need to separate it out. Just updated!

Reply:
yep, i find that having more useEffect blocks actually often helps with readability and structure!